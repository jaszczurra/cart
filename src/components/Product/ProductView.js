import React from 'react'
import { Field } from 'formik'
import PropTypes from 'prop-types'

import { CounterField } from '../Counter/CounterField'

import { convertPrice } from '../../utils/helpers'

const ProductView = ({ id, product }) => {
  const { name, price, min, max, isBlocked } = product
  const convertedPrice = convertPrice(price)

  return (
    <li className='row'>
      <span>{`${name}, cena: ${convertedPrice}z≈Ç`}</span>
      <Field
        name={`products[${id}].quantity`}
        component={CounterField}
        min={min}
        max={max}
        isBlocked={isBlocked}
      />
    </li>
  )
}

ProductView.propTypes = {
  id: PropTypes.number.isRequired,
  product: PropTypes.object.isRequired,
}

export { ProductView }
