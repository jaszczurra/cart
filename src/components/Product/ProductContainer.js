import React, { useEffect } from 'react'
import isEmpty from 'lodash/isEmpty'
import PropTypes from 'prop-types'

import { ProductView } from './ProductView'

const ProductContainer = ({ id, product, validationData, setFieldValue }) => {

  useEffect(() => {
    validationData && validationData.isError && setFieldValue(`products[${id}].quantity`, product.min)
  }, [validationData])

  if (isEmpty(product)) return null
  return (
    <ProductView
      id={id}
      product={product}
    />
  )
}

ProductContainer.propTypes = {
  id: PropTypes.number.isRequired,
  product: PropTypes.object,
  validationData: PropTypes.object,
  setFieldValue: PropTypes.func,
}

export { ProductContainer }
