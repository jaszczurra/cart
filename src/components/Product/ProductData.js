import React, { useEffect, useCallback } from 'react'
import debounce from 'lodash/debounce'
import isEmpty from 'lodash/isEmpty'
import get from 'lodash/get'
import PropTypes from 'prop-types'

import { useFetch } from '../../hooks/useFetch'

import { ProductContainer } from './ProductContainer'

const ProductData = ({ id, product, values, setFieldValue }) => {
  const { data: validationData, callFetch } = useFetch()

  const value = get(values, `products[${id}].quantity`)

  const validateProduct = useCallback(debounce((pid, quantity) => {
    callFetch({
      url: '/api/product/check',
      payload: {
        pid,
        quantity
      }
    })
  }, 300), [])

  useEffect(() => {
    !isEmpty(product) && value && validateProduct(product.pid, value)
  }, [value])

  return (
    <ProductContainer
      id={id}
      product={product}
      validationData={validationData}
      setFieldValue={setFieldValue}
    />
  )
}

ProductData.propTypes = {
  id: PropTypes.number.isRequired,
  product: PropTypes.object,
  value: PropTypes.object,
  setFieldValue: PropTypes.func,
}

export { ProductData }
