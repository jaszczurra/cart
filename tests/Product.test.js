import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { mount } from 'enzyme'

import { ProductContainer } from '../src/components/Product/ProductContainer'
import { ProductView } from '../src/components/Product/ProductView'

const mockedProduct = {
  pid: '8e5e1248-c799-4937-9acc-2b3ab0e034ff',
  name: 'Patelnia',
  max: 10,
  min: 1,
  price: '89.99',
}

describe('Product', () => {

  it('Renders nothing if no product', () => {
    const component = mount(<ProductContainer id={0} />)
    expect(component.children().length).toBe(0)
  })

  it('Snapshot', () => {
    const renderer = new ShallowRenderer()
    const snapshot = renderer.render(<ProductView
      id={0}
      product={mockedProduct}
    />)

    expect(snapshot).toMatchSnapshot()
  })
})
