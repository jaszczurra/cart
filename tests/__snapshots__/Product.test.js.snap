// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Snapshot 1`] = `
<li
  className="row"
>
  <span>
    Patelnia, cena: 89,99z≈Ç
  </span>
  <Field
    component={[Function]}
    max={10}
    min={1}
    name="products[0].quantity"
  />
</li>
`;
